src/wp-includes/option.php

```diff
@@ -253,6 +253,118 @@ function get_option( $option, $default_value = false ) {
 	return apply_filters( "option_{$option}", maybe_unserialize( $value ), $option );
 }
 
+/**
+ * Primes specific options into the cache with a single database query.
+ *
+ * Only options that do not already exist in cache will be primed.
+ *
+ * @since 6.4.0
+ *
+ * @global wpdb $wpdb WordPress database abstraction object.
+ *
+ * @param array $options An array of option names to be primed.
+ */
+function prime_options( $options ) {
+	$alloptions     = wp_load_alloptions();
+	$cached_options = wp_cache_get_multiple( $options, 'options' );
+
+	// Filter options that are not in the cache.
+	$options_to_prime = array();
+	foreach ( $options as $option ) {
+		if ( ( ! isset( $cached_options[ $option ] ) || ! $cached_options[ $option ] ) && ! isset( $alloptions[ $option ] ) ) {
+			$options_to_prime[] = $option;
+		}
+	}
+
+	// Bail early if there are no options to be primed.
+	if ( empty( $options_to_prime ) ) {
+		return;
+	}
+
+	global $wpdb;
+	$results = $wpdb->get_results(
+		$wpdb->prepare(
+			sprintf(
+				"SELECT option_name, option_value FROM $wpdb->options WHERE option_name IN (%s)",
+				implode( ',', array_fill( 0, count( $options_to_prime ), '%s' ) )
+			),
+			$options_to_prime
+		)
+	);
+
+	$options_found = array();
+	foreach ( $results as $result ) {
+		$options_found[ $result->option_name ] = maybe_unserialize( $result->option_value );
+	}
+	wp_cache_set_multiple( $options_found, 'options' );
+
+	// If all options were found, no need to update `notoptions` cache.
+	if ( count( $options_found ) === count( $options_to_prime ) ) {
+		return;
+	}
+
+	$options_not_found = array_diff( $options_to_prime, array_keys( $options_found ) );
+
+	$notoptions = wp_cache_get( 'notoptions', 'options' );
+
+	// Make a copy to check later if any change was made.
+	$new_notoptions = $notoptions;
+
+	// Add the options that were not found to the cache.
+	if ( ! is_array( $new_notoptions ) ) {
+		$new_notoptions = array();
+	}
+
+	foreach ( $options_not_found as $option_name ) {
+		if ( ! isset( $new_notoptions[ $option_name ] ) ) {
+			$new_notoptions[ $option_name ] = true;
+		}
+	}
+
+	// Only update the cache if it was modified.
+	if ( $new_notoptions !== $notoptions ) {
+		wp_cache_set( 'notoptions', $new_notoptions, 'options' );
+	}
```

Comment:
Check to see if `$options_not_found` is empty before handling `notoptions` and you can skip accessing the cache and simplify things a bit.

```suggestion
	$options_not_found = array_diff( $options_to_prime, array_keys( $options_found ) );

	// If any options are not found, make sure they get added to the notoptions cache.
	if ( ! empty( $options_not_found ) ) {
		$notoptions = wp_cache_get( 'notoptions', 'options' );

		// Add the options that were not found to the cache.
		if ( ! is_array( $notoptions ) ) {
			$notoptions = array();
		}

		foreach ( $options_not_found as $option_name ) {
			if ( ! isset( $notoptions[ $option_name ] ) ) {
				$notoptions[ $option_name ] = true;
			}
		}

		wp_cache_set( 'notoptions', $notoptions, 'options' );
	}
```

Reply:
@joemcgill This is unnecessary as it would duplicate what is already achieved right above with the `count` comparison.

Reply:
Ah, I missed the early return. Thanks for pointing that out. I think the current approach is fine. One nitpick (not a blocker) is that you could set a variable for whether an option has been added to the `$notoptions` array and avoid duplicating the whole array. Something like:

```
		$update_notoptions = false;

		foreach ( $options_not_found as $option_name ) {
			if ( ! isset( $notoptions[ $option_name ] ) ) {
				$notoptions[ $option_name ] = true;
				$update_notoptions = true;
			}
		}

		if ( $update_notoptions ) {
			wp_cache_set( 'notoptions', $notoptions, 'options' );
		}
```

Reply:
@joemcgill Makes sense, that's certainly cleaner. Updated in https://github.com/WordPress/wordpress-develop/pull/4970/commits/9682336fcd8a03ac4e11f310849a99c298a8c615.